Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift

```diff
@@ -15,38 +15,92 @@ import Dispatch
+        /// always override any other kind of package reference and branch-based reference will override
+        /// version-based reference.
+        let overriddenPackages: [PackageReference: (version: BoundVersion, products: ProductFilter)]
+
+        /// A collection of all known incompatibilities matched to the packages they
+        /// refer to. This means an incompatibility can occur several times.
+        public private(set) var incompatibilities: [DependencyResolutionNode: [Incompatibility]] = [:]
+
+        /// The current best guess for a solution satisfying all requirements.
+        public private(set) var solution: PartialSolution
+        
+        private let lock = Lock()
+
+        init(root: DependencyResolutionNode,
+             overriddenPackages: [PackageReference: (version: BoundVersion, products: ProductFilter)] = [:],
+             solution: PartialSolution = PartialSolution()) {
+            self.root = root
+            self.overriddenPackages = overriddenPackages
+            self.solution = solution
+        }
 
-    /// Find all incompatibilities containing a positive term for a given package.
-    public func positiveIncompatibilities(for node: DependencyResolutionNode) -> [Incompatibility]? {
-        guard let all = incompatibilities[node] else {
-            return nil
+        func addIncompatibility(_ incompatibility: Incompatibility, at location: LogLocation) {
+            self.lock.withLock {
+                //log("incompat: \(incompatibility) \(location)")
+                for package in incompatibility.terms.map({ $0.node }) {
+                    if let incompats = self.incompatibilities[package] {
+                        if !incompats.contains(incompatibility) {
+                            self.incompatibilities[package]!.append(incompatibility)
+                        }
+                    } else {
+                        self.incompatibilities[package] = [incompatibility]
+                    }
+                }
+            }
         }
-        return all.filter {
-            $0.terms.first { $0.node == node }!.isPositive
+
+        /// Find all incompatibilities containing a positive term for a given package.
+        func positiveIncompatibilities(for node: DependencyResolutionNode) -> [Incompatibility]? {
+            self.lock.withLock {
+                guard let all = self.incompatibilities[node] else {
+                    return nil
+                }
+                return all.filter {
+                    $0.terms.first { $0.node == node }!.isPositive
+                }
+            }
         }
-    }
 
-    /// The root package reference.
-    private(set) var root: DependencyResolutionNode?
+        func decide(_ node: DependencyResolutionNode, at version: Version) {
+            let term = Term(node, .exact(version))
+            // FIXME: Shouldn't we check this _before_ making a decision?
```

Comment:
Existing logic: is this FIXME still needed? `isValidDecision` is indeed called before `decide` (L82) no?

Reply:
trying to minimize changes for now, we can revisit in later PR